apps/backend/src/lib/googleOAuth.ts

```diff
@@ -0,0 +1,59 @@
+import { URL } from 'url';
+import { OAuth2Client } from 'google-auth-library';
+import { backendEnvSchema } from 'src/env';
+import { TRPCError } from '@trpc/server';
+
+export class GoogleOAuth2ClientsManager {
+    /**
+     * Map of redirect origins to OAuth2Clients.
+     */
+    private clients: Map<string, OAuth2Client> = new Map();
+    private readonly clientId: string;
+    private readonly clientSecret: string;
+    private readonly allowDynamicRedirects: boolean;
+    private readonly redirectOrigin: string;
+
+    constructor(_clientId: string, _clientSecret: string) {
+        this.clientId = _clientId;
+        this.clientSecret = _clientSecret;
+
+        const { STAGE, GOOGLE_REDIRECT_URI } = backendEnvSchema.parse(process.env);
+        this.allowDynamicRedirects = STAGE !== 'prod';
+        this.redirectOrigin = GOOGLE_REDIRECT_URI;
+    }
+
+    getClient(redirectOrigin: string): OAuth2Client {
```

Comment:
A docstring here, including what the important conditions are, would be good, too